### PR TITLE
Use async cosmos library and upgrade to 2.6.3

### PR DESCRIPTION
**Pull request by ankagrawal**
This changes replaces the cosmos documentdb library with async cosmosdb library.

### CODE REVIEW

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java [hunk 1]
```diff
@@ -166,96 +161,34 @@ void testConnectivity() {
...
 
-  /**
-   * Run the supplied DocumentClient action. If CosmosDB returns status 429 (TOO_MANY_REQUESTS),
-   * retry after the requested wait period, up to the configured retry limit.
-   * @param operation the DocumentClient resource operation to execute, wrapped in a {@link Callable}.
-   * @param timer the {@link Timer} to use to record execution time (excluding waiting).
-   * @return the {@link ResourceResponse} returned by the operation, if successful.
-   * @throws DocumentClientException if Cosmos returns a different error status, or if the retry limit is reached.
-   */
-  private ResourceResponse<Document> retryOperationWithThrottling(Callable<ResourceResponse<Document>> operation,
```
Comment (ankagrawal) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java]: Note that some default retry for http 429 happens implicity in the async SDK (As per [https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/cosmos-db/performance-tips-async-java.md] ). So I decided to do away with our own throttling logic.

_

> The SDKs all implicitly catch this response, respect the server-specified retry-after header, and retry the request. Unless your account is being accessed concurrently by multiple clients, the next retry will succeed.
> 
> If you have more than one client cumulatively operating consistently above the request rate, the default retry count currently set to 9 internally by the client may not suffice; in this case, the client throws a DocumentClientException with status code 429 to the application. The default retry count can be changed by using setRetryOptions on the ConnectionPolicy instance. By default, the DocumentClientException with status code 429 is returned after a cumulative wait time of 30 seconds if the request continues to operate above the request rate. This occurs even when the current retry count is less than the max retry count, be it the default of 9 or a user-defined value.

_

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 2]
```diff
@@ -127,16 +125,18 @@
...
+      connectionPolicy.setProxy(cloudConfig.vcrProxyHost, cloudConfig.vcrProxyPort);
     }
     // TODO: test option to set connectionPolicy.setEnableEndpointDiscovery(false);
-    documentClient = new DocumentClient(azureCloudConfig.cosmosEndpoint, azureCloudConfig.cosmosKey, connectionPolicy,
-        ConsistencyLevel.Session);
-    cosmosDataAccessor = new CosmosDataAccessor(documentClient, azureCloudConfig, azureMetrics);
+    asyncDocumentClient = new AsyncDocumentClient.Builder().withServiceEndpoint(azureCloudConfig.cosmosEndpoint)
+        .withMasterKeyOrResourceToken(azureCloudConfig.cosmosKey)
+        .withConnectionPolicy(connectionPolicy)
+        .withConsistencyLevel(ConsistencyLevel.Session)
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: We probably need to change this but fine to keep it for now.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 3]
```diff
@@ -127,16 +125,18 @@
     ConnectionPolicy connectionPolicy = new ConnectionPolicy();
     if (azureCloudConfig.cosmosDirectHttps) {
       logger.info("Using CosmosDB DirectHttps connection mode");
-      connectionPolicy.setConnectionMode(ConnectionMode.DirectHttps);
+      connectionPolicy.setConnectionMode(ConnectionMode.Direct);
     }
     if (cloudConfig.vcrProxyHost != null) {
-      connectionPolicy.setProxy(new HttpHost(cloudConfig.vcrProxyHost, cloudConfig.vcrProxyPort));
-      connectionPolicy.setHandleServiceUnavailableFromProxy(true);
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: Did this method go away?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 4]
```diff
@@ -309,15 +306,14 @@ public boolean updateBlobExpiration(BlobId blobId, long expirationTime) throws C
...
+    return metadataList.stream().collect(Collectors.toMap(CloudBlobMetadata::getId, Function.identity()));
   }
 
   private List<CloudBlobMetadata> getBlobMetadataChunked(List<BlobId> blobIds) throws CloudStorageException {
     if (blobIds.isEmpty() || blobIds.size() > ID_QUERY_BATCH_SIZE) {
       throw new IllegalArgumentException("Invalid input list size: " + blobIds.size());
     }
-    String quotedBlobIds =
-        String.join(",", blobIds.stream().map(s -> '"' + s.getID() + '"').collect(Collectors.toList()));
+    String quotedBlobIds = blobIds.stream().map(s -> '"' + s.getID() + '"').collect(Collectors.joining(","));
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: nice

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 5]
```diff
@@ -392,7 +388,7 @@ private void filterOutLastReadBlobs(List<CloudBlobMetadata> cloudBlobMetadataLis
    * @param fieldName The metadata field to modify.
    * @param value The new value.
    * @return {@code true} if the udpate succeeded, {@code false} if the metadata record was not found.
-   * @throws DocumentClientException
+   * @throws CloudStorageException
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: good catch.  please add "if the update fails."

File: ambry-cloud/src/test/java/com.github.ambry.cloud/azure/AzureCloudDestinationTest.java [hunk 6]
```diff
@@ -502,62 +501,36 @@ public void testAzureConnection() throws Exception {
     }
   }
 
-  /** Test initializing AzureCloudDestination with a proxy */
-  @Test
-  public void testProxy() throws Exception {
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com.github.ambry.cloud/azure/AzureCloudDestinationTest.java]: Why was this deleted?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 7]
```diff
@@ -127,16 +125,18 @@
...
-      connectionPolicy.setHandleServiceUnavailableFromProxy(true);
+      connectionPolicy.setProxy(cloudConfig.vcrProxyHost, cloudConfig.vcrProxyPort);
     }
     // TODO: test option to set connectionPolicy.setEnableEndpointDiscovery(false);
-    documentClient = new DocumentClient(azureCloudConfig.cosmosEndpoint, azureCloudConfig.cosmosKey, connectionPolicy,
-        ConsistencyLevel.Session);
-    cosmosDataAccessor = new CosmosDataAccessor(documentClient, azureCloudConfig, azureMetrics);
+    asyncDocumentClient = new AsyncDocumentClient.Builder().withServiceEndpoint(azureCloudConfig.cosmosEndpoint)
+        .withMasterKeyOrResourceToken(azureCloudConfig.cosmosKey)
+        .withConnectionPolicy(connectionPolicy)
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: should we make the number of retries configurable?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 8]
```diff
@@ -359,6 +363,14 @@ public boolean updateBlobExpiration(BlobId blobId, long expirationTime) throws C
     }
   }
 
+  /**
+   * Getter for {@link AsyncDocumentClient} object.
+   * @return {@link AsyncDocumentClient} object.
+   */
+  public AsyncDocumentClient getAsyncDocumentClient() {
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: This can be package private, no?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 9]
```diff
@@ -309,15 +310,14 @@ public boolean updateBlobExpiration(BlobId blobId, long expirationTime) throws C
...
   private List<CloudBlobMetadata> getBlobMetadataChunked(List<BlobId> blobIds) throws CloudStorageException {
     if (blobIds.isEmpty() || blobIds.size() > ID_QUERY_BATCH_SIZE) {
       throw new IllegalArgumentException("Invalid input list size: " + blobIds.size());
     }
-    String quotedBlobIds =
-        String.join(",", blobIds.stream().map(s -> '"' + s.getID() + '"').collect(Collectors.toList()));
+    String quotedBlobIds = blobIds.stream().map(s -> '"' + s.getID() + '"').collect(Collectors.joining(","));
     String query = String.format(BATCH_ID_QUERY_TEMPLATE, quotedBlobIds);
     String partitionPath = blobIds.get(0).getPartition().toPathString();
     try {
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: This may be an action item for the future, but we should consider if we can get rid of a lot of the string concat/formatting when running these queries.

There is an alternate constructor for `SqlQuerySpec` that is used in `getDeadBlobs`: `SqlQuerySpec(String queryText, SqlParameterCollection parameters)`

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java [hunk 10]
```diff
@@ -16,102 +16,96 @@
...
-  public CosmosDataAccessor(DocumentClient documentClient, AzureCloudConfig azureCloudConfig,
+  CosmosDataAccessor(AsyncDocumentClient asyncDocumentClient, AzureCloudConfig azureCloudConfig,
       AzureMetrics azureMetrics) {
-    this(documentClient, azureCloudConfig.cosmosCollectionLink, azureCloudConfig.cosmosMaxRetries, azureMetrics);
+    this(asyncDocumentClient, azureCloudConfig.cosmosCollectionLink, azureMetrics);
   }
 
   /** Test constructor */
-  public CosmosDataAccessor(DocumentClient documentClient, String cosmosCollectionLink, int maxRetries,
+  public CosmosDataAccessor(AsyncDocumentClient asyncDocumentClient, String cosmosCollectionLink,
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java]: why does the test constructor need to be public?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java [hunk 11]
```diff
@@ -166,96 +161,34 @@ void testConnectivity() {
...
-
   /**
    * Utility method to call a Cosmos method and extract any nested DocumentClientException.
    * @param action the action to call.
    * @return the result of the action.
-   * @throws Exception
+   * @throws DocumentClientException
    */
-  private Object executeCosmosAction(Callable<? extends Object> action) throws Exception {
+  private ResourceResponse<Document> executeCosmosAction(Callable<? extends Object> action, Timer timer)
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java]: would making the parameter a `Callable<? extends ResourceResponse<? extends Document>>`, or `Callable<ResourceResponse<Document>>` work?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java [hunk 12]
```diff
@@ -166,96 +161,34 @@ void testConnectivity() {
...
-    } catch (DocumentClientException dex) {
-      throw dex;
-    } catch (IllegalStateException ex) {
-      if (ex.getCause() instanceof DocumentClientException) {
-        throw (DocumentClientException) ex.getCause();
-      } else {
-        throw ex;
+      Timer.Context operationTimer = timer.time();
+      resourceResponse = (ResourceResponse<Document>) action.call();
+      operationTimer.stop();
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java]: put the stop in a finally clause?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java [hunk 13]
```diff
@@ -166,96 +161,34 @@ void testConnectivity() {
...
+    } catch (RuntimeException rex) {
+      if (rex.getCause() instanceof DocumentClientException) {
+        throw (DocumentClientException) rex.getCause();
       }
+      throw (RuntimeException) rex.getCause();
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java]: this else case should just `throw rex`, right? since the cause may not necessarily be a `RuntimeException`

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java (hunk 9 above)
Comment (cgtz) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: Nevermind, I found that the client doesn't support making list `SqlParameter` that could be used for in clauses: https://github.com/Azure/azure-cosmosdb-java/issues/60

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java (hunk 8 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: done.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java (hunk 10 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java]: fixed.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java (hunk 11 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java]: done.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java (hunk 12 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java]: done.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java (hunk 13 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosDataAccessor.java]: True. I dont know why I did that. fixed.

File: ambry-cloud/src/test/java/com.github.ambry.cloud/azure/AzureTestUtils.java [hunk 14]
```diff
@@ -43,10 +53,52 @@ public static Document createDocumentFromCloudBlobMetadata(CloudBlobMetadata clo
...
+    when(mockResourceResponse.getResource()).thenReturn(metadataDoc);
+    return mockResponse;
+  }
+
+  /**
+   * Utility to mock the query call chain of {@link AsyncDocumentClient} such that query returns {@code documentList}.
+   * @param documentList {@link List <Document>} of documents to return from mocked call.
+   * @param mockResponse {@link Observable} mocked response.
+   */
+  static void mockObservableForQuery(List<Document> documentList, Observable<FeedResponse<Document>> mockResponse) {
```
Comment (lightningrob) [inline ambry-cloud/src/test/java/com.github.ambry.cloud/azure/AzureTestUtils.java]: Why doesn't this method create and return the mockResponse like the others?
